Add slotFills for header items

octaedro opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR adds slotFills to the product editor header.

<img width="529" alt="Screenshot 2024-01-16 at 15 21 44" src="https://github.com/woocommerce/woocommerce/assets/1314156/1beaa5b3-21b7-445f-b78b-1403e6781ee3">

### Extensibility

**PluginHeaderItemPopover**
```
<PluginHeaderItemPopover
    label={}
    icon={}
>
    <div>Popover content</div>
</PluginHeaderItemPopover>
```

**PluginHeaderItemModal**

```
<PluginHeaderItemModal
    label={}
    icon={}
    title={ 'My title' }
>
  <div>Modal content</div>
</PluginHeaderItemModal>
```

These SlotFills allow for the header to be extended by extensions with popovers and modals. We are using `PinnedItems.Slot` under the hood.

This PR does what the issue requests, but [I left in the issue a few extra questions](https://github.com/woocommerce/woocommerce/issues/42805#issuecomment-1893296138) regarding a follow-up for this functionality. 

@mattsherman, in the issue was suggested to use `PinnedItems`, but [it doesn't order the `fills`](https://github.com/WordPress/gutenberg/blob/trunk/packages/interface/src/components/pinned-items/index.js) items. This is why I removed the prop from the components. 

Closes #42805.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure the feature `New product editor` is enabled under `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`.
2. To make testing easier, I created [this plugin](https://github.com/woocommerce/woocommerce/files/13996806/myplugin.zip). Please download it, install and activate it.
3. Go to `Products > Add New`.
4. Verify the items were added to the product header. 

![Screen Capture on 2024-01-16 at 15-35-27](https://github.com/woocommerce/woocommerce/assets/1314156/876d49e9-a080-44df-9b6f-23adc4db2051)

5. Click them and verify they work as expected.

The testing plugin will add the following components:
```
<PluginHeaderItemPopover
	icon={ plugins }
	label={ 'My plugin' }
>
	<div className="my-plugin-class-name">
		<p>
			Popover without button or callback.
		</p>
	</div>
</PluginHeaderItemPopover>
<PluginHeaderItemPopover
	icon={ comment }
	label={ 'Comment extension' }
>
	{ ( { isVisible, setVisible }: { isVisible: boolean, setVisible: ( visible: boolean ) => void } ) => (
		<div className="my-plugin-class-name">
			<p>
				Popover with close button.
			</p>
			<Button
				onClick={ ()=> {
					console.log( `The popover is: ${ isVisible ? 'visible' : 'hidden' }, let's hide it!`)
					setVisible( false );
				} }
				variant="primary"
			>
				Close popover!
			</Button>
		</div>
	) }
</PluginHeaderItemPopover>
<PluginHeaderItemModal
	icon={ arrowDown }
	label={ 'Modal label' }
	title={ 'Title' }
>
	{ ( { isOpen, setOpen } ) => (
		<div className="my-plugin-class-name">
			<p>
				Modal with close button.
			</p>
			<Button
				onClick={ ()=> {
					console.log( `The modal is: ${ isOpen ? 'open' : 'hidden' }, let's close it!`)
					setOpen( false );
				} }
				variant="primary"
			>
				Close modal!
			</Button>
		</div>
	) }
</PluginHeaderItemModal>
<PluginHeaderItemModal
	icon={ info }
	label={ 'Modal label' }
	title={ 'Title' }
>
	<div className="my-plugin-class-name">
		<p>
			Modal without button or callback.
		</p>
	</div>
</PluginHeaderItemModal>
```

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

octaedro committed:
Add slotFills for header items

octaedro committed:
Refactor plugin-header-items

octaedro committed:
Add changelog

octaedro committed:
Remove order prop

octaedro committed:
Fix lint

mattsherman commented on packages/js/product-editor/src/constants.ts:
```diff
@@ -55,6 +55,8 @@ export const PRODUCT_SCHEDULED_SALE_SLUG = 'product-scheduled-sale';
 
 export const TRACKS_SOURCE = 'product-block-editor-v1';
 
+export const PINNED_ITEMS_SCOPE = 'woocommerce/product-editor';
```
The name of the constant should either be more specific, to mention that it is for the header pinned items.

octaedro replied:
Done in commit [b05d244](https://github.com/woocommerce/woocommerce/pull/43687/commits/b05d244f35c5e356b45bf572dbca2f42db3d96ec) and [b1b7dea](https://github.com/woocommerce/woocommerce/pull/43687/commits/b1b7deaa2a10ad7a234c6bf94037177864d6e737)

mattsherman commented on packages/js/product-editor/src/components/header/plugin-header-items/types.tsx:
```diff
@@ -0,0 +1,12 @@
+export type PluginHeaderItemModalProps = {
+	children?: React.ReactNode;
+	icon: React.ReactNode;
```
`icon` should be optional, and default to a generic plugin icon if not provided.

octaedro replied:
I like the idea. We need a design for that (for the default icon we will use when it's not provided). Let's ask @j111q.

I wouldn't make the `icon` optional without having a generic one to use when there is no icon. Maybe we can use the `plugins` icon, what do you think?

<img width="230" alt="Screenshot 2024-01-20 at 11 57 59" src="https://github.com/woocommerce/woocommerce/assets/1314156/1b1597aa-ec7e-4a87-b25f-7fa614fe0e56">

mattsherman replied:
I think the `plugins` icon makes sense. That's what I was referring to. Gutenberg uses that too.

mattsherman replied:
Keep in mind that even marking `icon` as required goes out the door if the extension dev is not using TypeScript. So, we should always gracefully handle fallbacks in cases like this.

octaedro replied:
Done in commits [514ea78](https://github.com/woocommerce/woocommerce/pull/43687/commits/514ea78e15d89f43c60c9aa04944755000d69794) and [c133e70](https://github.com/woocommerce/woocommerce/pull/43687/commits/c133e7086f5de88ddfbdd97de3474e599c484d7f)

mattsherman commented on packages/js/product-editor/src/components/header/plugin-header-items/types.tsx:
```diff
@@ -0,0 +1,12 @@
+
+export type PluginHeaderItemPopoverProps = {
+	children?: React.ReactNode;
+	icon: React.ReactNode;
```
`icon` should be optional, and default to a generic plugin icon if not provided.

mattsherman commented on packages/js/product-editor/src/components/header/plugin-header-items/plugin-header-item-modal.tsx:
```diff
@@ -0,0 +1,44 @@
+						title={ title }
+						onRequestClose={ () => setOpen( false ) }
+					>
+						{ children }
```
We should pass a callback to the children to allow the extension to close the modal (for example, after a button is clicked on the modal).

octaedro replied:
Done in commit [3969d4a](https://github.com/woocommerce/woocommerce/pull/43687/commits/3969d4ab9dae82f9ebe574171e3e27fb8e074192)

mattsherman reviewed (changes requested):
Test and overall is looking very nice! Very easy to add header items.

Some issues:

- If all the header items can't be shown, only the first one is shown, with no way to access the remaining header items
- No way to close a modal from the modal
- Icon should default to generic plugin icon if no icon is provided

Test code used:

```js
function generateRandomString(length) {
    let result = '';
    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    const charactersLength = characters.length;
    
    for (let i = 0; i < length; i++) {
      const randomIndex = Math.floor(Math.random() * charactersLength);
      result += characters.charAt(randomIndex);
    }
    
    return result;
}

wp.plugins.registerPlugin( 'example-variation-update-menu', {
    scope: 'woocommerce-product-block-editor',
    render: function() {
        const postType = wp.element.useContext( wc.productEditor.PostTypeContext );
        const [ id ] = wp.coreData.useEntityProp( 'postType', postType, 'id' );
        const [ name, setName ] = wc.productEditor.__experimentalUseProductEntityProp( 'name' );

        return wp.element.createElement(
            wp.element.Fragment,
            {},
            
            // Popover
            wp.element.createElement(
                wc.productEditor.PluginHeaderItemPopover,
                {
                    label: 'Plugin Header Item Popover',
                },

                wp.element.createElement(
                    'div',
                    {
                        style: {
                            padding: '32px',
                            width: '300px',
                        }
                    },

                    wp.element.createElement(
                        'div',
                        {},
                        `Post type: ${ postType }`,
                    ),

                    wp.element.createElement(
                        'div',
                        {},
                        `Id: ${ id }`,
                    ),

                    wp.element.createElement(
                        'div',
                        {},
                        `Name: ${ name }`,
                    ),
                ),
            ),

            // Modal
            wp.element.createElement(
                wc.productEditor.PluginHeaderItemModal,
                {
                    label: 'Plugin Header Item Modal',
                    title: 'Plugin Header Item Modal',
                },

                wp.element.createElement(
                    'div',
                    {},
                    `Post type: ${ postType }`,
                ),

                wp.element.createElement(
                    'div',
                    {},
                    `Id: ${ id }`,
                ),

                wp.element.createElement(
                    'div',
                    {},
                    `Name: ${ name }`,
                ),

                wp.element.createElement(
                    wp.components.Button,
                    {
                        variant: 'primary',
                        onClick: () => { setName( generateRandomString( 10 ) ) },
                    },
                    'Set Name to Random String',
                ), 
            ),

            wp.element.createElement(
                wc.productEditor.PluginHeaderItemPopover,
                {
                    label: 'Plugin Header Item Popover 3',
                },

                wp.element.createElement(
                    'div',
                    {
                        style: {
                            padding: '32px',
                            width: '300px',
                        }
                    },
                    
                    'Plugin Header Item Popover 3',
                ),
            ),

            // Popover
            wp.element.createElement(
                wc.productEditor.PluginHeaderItemPopover,
                {
                    label: 'Plugin Header Item Popover 4',
                },

                wp.element.createElement(
                    'div',
                    {
                        style: {
                            padding: '32px',
                            width: '300px',
                        }
                    },
                    
                    'Plugin Header Item Popover 4',
                ),
            ),

            // Popover
            wp.element.createElement(
                wc.productEditor.PluginHeaderItemPopover,
                {
                    label: 'Plugin Header Item Popover 5',
                },

                wp.element.createElement(
                    'div',
                    {
                        style: {
                            padding: '32px',
                            width: '300px',
                        }
                    },
                    
                    'Plugin Header Item Popover 5',
                ),
            ),

            // Popover
            wp.element.createElement(
                wc.productEditor.PluginHeaderItemPopover,
                {
                    label: 'Plugin Header Item Popover 6',
                },

                wp.element.createElement(
                    'div',
                    {
                        style: {
                            padding: '32px',
                            width: '300px',
                        }
                    },
                    
                    'Plugin Header Item Popover 6',
                ),
            ),
        );
    }
} );
```

octaedro committed:
Change constant name

octaedro committed:
Rename constant

octaedro committed:
Add callback to fills

octaedro commented:
Thank you for reviewing this PR @mattsherman! I addressed your suggestions and updated [the testing plugin](https://github.com/woocommerce/woocommerce/files/13996806/myplugin.zip) to test the new changes. Now shows in the browser's console the `isOpen` prop for the modal and popover, and adds a button to close them.

Regarding this point:

> If all the header items can't be shown, only the first one is shown, with no way to access the remaining header items

One of the extensibility guideline points says: `Each top bar extension has its unique icon in the top navigation bar. Note that when the number of extensions exceeds 4, they're truncated in a dropdown menu`. We might use the same approach for the icons on small screens.

Since the design is not very clear for this. What do you think about hiding the icons on small screens ( < 600px ) and tackling these icon behaviors in a follow-up PR?

mattsherman commented:
> What do you think about hiding the icons on small screens ( < 600px ) and tackling these icon behaviors in a follow-up PR?

mattsherman reviewed (changes requested):
Thanks for adding the ability to close the modal! Works great.

Once you have a default for the `icon` in, I'll re-review and can approve this PR!

octaedro committed:
Make icon not required

octaedro committed:
Add fallback plugin

octaedro commented:
> Can't we just do what the guidelines say and put them in a dropdown menu?

I would keep the scope of this PR as it is and tackle that in a follow-up PR. Also, we already have in the header a `vertical ellipsis` (moreVertical) icon, so I think it would make sense to ask @j111q which icon we should use in that case. Maybe we can repeat the `plugins` icon but I think it would be good to confirm.

octaedro commented:
Hi @mattsherman,
I added a default icon and made it not required. Could you take another look at this PR?

mattsherman commented:
> > Can't we just do what the guidelines say and put them in a dropdown menu?
> 
> I would keep the scope of this PR as it is and tackle that in a follow-up PR.

I'm okay with that. It will keep us moving, while we figure out the UX details, which is nice! 

>  Also, we already have in the header a `vertical ellipsis` (moreVertical) icon, so I think it would make sense to ask @j111q which icon we should use in that case. Maybe we can repeat the `plugins` icon but I think it would be good to confirm.

True, true... so, we should do what the post editor does and put them in the `...` menu. @j111q , can you please review how this is handled in the post editor and see if we can follow suit for our header items? I think it is good to be consistent between our experience and WP core.

mattsherman reviewed (approved):
Thanks for adding the default icon. Much better.

I'm okay with the overflow handling (when all the icons can't be shown in the header) being handled in a follow up. Will you handle that, or do you have other priorities presently?

j111q commented:
> True, true... so, we should do what the post editor does and put them in the ... menu. @j111q , can you please review how this is handled in the post editor and see if we can follow suit for our header items? I think it is good to be consistent between our experience and WP core.

Will do and agree about the consistency. 👀

octaedro commented:
Hi @mattsherman 

> I'm okay with the overflow handling (when all the icons can't be shown in the header) being handled in a follow-up. Will you handle that, or do you have other priorities presently?

I have other priorities for this week, but I'll be happy to create an issue to tackle that.
